key.go

```diff
@@ -0,0 +1,35 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package rate
+
+import (
+	"bytes"
+	"sync"
+)
+
+var keyBuilderPool sync.Pool
+
+type builder struct {
+	bytes.Buffer
+}
+
+func getKey(parts ...string) string {
```

Comment:
perhaps optimize for the cases where len(parts) is 0 or 1?

Reply:
The way this gets used `len(parts)` will either be 3 or 4. If this was for external use it would make sense to optimize for 0 or 1.

Reply:
any reason to not add a constraint on len then?

Reply:
I think it *could* be possible to support other lengths in the future, so it doesn't seem like an error to pass in other lengths to this function. And I think if there is a constraint, it would be enforced by the caller, where the distinction between 3 or 4 parts would be meaningful.